More context for errors

kolyshkin opened the pull request:
Some functions that moby uses return "raw" errors from the kernel. Wrap those to provide more error context to the caller. Please see individual commit descriptions for details.

kolyshkin committed:
pkg/system/stat_unix: wrap errors in PathError

syscall.Stat (and Lstat), unlike functions from os pkg,
return "raw" errors (like EPERM or EINVAL), and those are
propagated up the function call stack unchanged, and gets
logged and/or returned to the user as is.

Wrap those into os.PathError{} so the error message will
at least have function name and file name.

Note we use Capitalized function names to distinguish
between functions in os and ours.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
pkg/mount: wrap mount/umount errors

The errors returned from Mount and Unmount functions are raw
syscall.Errno errors (like EPERM or EINVAL), which provides
no context about what has happened and why.

Similar to os.PathError type, introduce mount.Error type
with some context. The error messages will now look like this:

> mount /tmp/mount-tests/source:/tmp/mount-tests/target, flags: 0x1001: operation not permitted

or

> mount tmpfs:/tmp/mount-test-source-516297835: operation not permitted

Before this patch, it was just

> operation not permitted

[v2: add Cause()]
[v3: rename MountError to Error, document Cause()]
[v4: fixes; audited all users]
[v5: make Error type private; changes after @cpuguy83 reviews]

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

AkihiroSuda commented on pkg/mount/mount.go:
```diff
@@ -2,12 +2,42 @@ package mount // import "github.com/docker/docker/pkg/mount"
+	source, target string
+	flags          uintptr
+	data           string
+	err            error
```
Could you implement causer https://godoc.org/github.com/pkg/errors#Cause

AkihiroSuda commented:
LGTM but a small nit.

Could you port this to containerd, because we are going to deprecate docker/pkg/mount pkg and use containerd

AkihiroSuda commented:
```
05:28:00 pkg/mount/mount.go:41:1:warning: exported method MountError.Cause should have comment or be unexported (golint)
05:28:00 pkg/mount/mount.go:13:6:warning: type name will be used as mount.MountError by other packages, and that stutters; consider calling this Error (golint)
```

thaJeztah commented:
Failures look legit (at least, same failures on janky, power, and z)

```
03:41:31 --- FAIL: TestSubtreePrivate (0.04s)
03:41:31     sharedsubtree_linux_test.go:65: %!v(PANIC=runtime error: invalid memory address or nil pointer dereference)
03:41:31 --- FAIL: TestSubtreeSharedSlave (0.06s)
03:41:31     sharedsubtree_linux_test.go:249: %!v(PANIC=runtime error: invalid memory address or nil pointer dereference)
03:41:31 --- FAIL: TestSubtreeUnbindable (0.02s)
03:41:31     sharedsubtree_linux_test.go:330: mount /tmp/mount-tests/source:/tmp/mount-tests/target, flags: 0x1000: invalid argument
03:41:31 FAIL

kolyshkin committed:
UnmountIpcMount: simplify

As standard mount.Unmount does what we need, let's use it.

In addition, this adds ignoring "not mounted" condition, which
was previously implemented (see PR#33329, commit cfa2591d3f26)
via a very expensive call to mount.Mounted().

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
aufs: get rid of mount()

The function is not needed as it's just a shallow wrapper around
unix.Mount().

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
I have now carefully audited all users of `pkg/mount` and made all the necessary changes (in most cases it is simplification of logging as the error already contains everything needed).

kolyshkin commented:
Failure in experimental CI is flaky test, `TestServiceWithDefaultAddressPoolInit`, https://github.com/moby/moby/pull/37836

Windows failure is unrelated

kolyshkin commented:
z failure is known flaky test, TestSwarmClusterRotateUnlockKey, see https://github.com/moby/moby/issues/33041

thaJeztah commented on pkg/mount/mount.go:
```diff
@@ -90,7 +126,7 @@ func ForceMount(device, target, mType, options string) error {
 func Unmount(target string) error {
 	err := unmount(target, mntDetach)
-	if err == syscall.EINVAL {
+	if err != nil && errors.Cause(err) == syscall.EINVAL {
```
Looks like we always want to ignore the `EINVAL` case (a similar check is in `RecursiveUnmount`); perhaps move this to `unmount()` and return `nil` if the error is due to `sys call.EINVAL`;

```patch
diff --git a/pkg/mount/mount.go b/pkg/mount/mount.go
index 12f968cbb8..3edd031acf 100644
--- a/pkg/mount/mount.go
+++ b/pkg/mount/mount.go
@@ -4,9 +4,7 @@ import (
 	"sort"
 	"strconv"
 	"strings"
-	"syscall"
 
-	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
@@ -125,12 +123,7 @@ func ForceMount(device, target, mType, options string) error {
 // Unmount lazily unmounts a filesystem on supported platforms, otherwise
 // does a normal unmount.
 func Unmount(target string) error {
-	err := unmount(target, mntDetach)
-	if err != nil && errors.Cause(err) == syscall.EINVAL {
-		// ignore "not mounted" error
-		err = nil
-	}
-	return err
+	return unmount(target, mntDetach)
 }
 
 // RecursiveUnmount unmounts the target and all mounts underneath, starting with
@@ -150,16 +143,6 @@ func RecursiveUnmount(target string) error {
 		logrus.Debugf("Trying to unmount %s", m.Mountpoint)
 		err = unmount(m.Mountpoint, mntDetach)
 		if err != nil {
-			// If the error is EINVAL either this whole package is wrong (invalid flags passed to unmount(2)) or this is
-			// not a mountpoint (which is ok in this case).
-			// Meanwhile calling `Mounted()` is very expensive.
-			//
-			// We've purposefully used `syscall.EINVAL` here instead of `unix.EINVAL` to avoid platform branching
-			// Since `EINVAL` is defined for both Windows and Linux in the `syscall` package (and other platforms),
-			//   this is nicer than defining a custom value that we can refer to in each platform file.
-			if errors.Cause(err) == syscall.EINVAL {
-				continue
-			}
 			if i == len(mounts)-1 {
 				if mounted, e := Mounted(m.Mountpoint); e != nil || mounted {
 					return err
diff --git a/pkg/mount/mounter_linux.go b/pkg/mount/mounter_linux.go
index bfa87a05ee..6f289e06d2 100644
--- a/pkg/mount/mounter_linux.go
+++ b/pkg/mount/mounter_linux.go
@@ -1,6 +1,9 @@
 package mount // import "github.com/docker/docker/pkg/mount"
 
 import (
+	"syscall"
+
+	"github.com/pkg/errors"
 	"golang.org/x/sys/unix"
 )
 
@@ -77,6 +80,16 @@ func unmount(target string, flag int) error {
 	if err == nil {
 		return nil
 	}
+	// If the error is EINVAL either this whole package is wrong (invalid flags passed to unmount(2)) or this is
+	// not a mountpoint (which is ok in this case).
+	// Meanwhile calling `Mounted()` is very expensive.
+	//
+	// We've purposefully used `syscall.EINVAL` here instead of `unix.EINVAL` to avoid platform branching
+	// Since `EINVAL` is defined for both Windows and Linux in the `syscall` package (and other platforms),
+	//   this is nicer than defining a custom value that we can refer to in each platform file.
+	if errors.Cause(err) == syscall.EINVAL {
+		return nil
+	}
 	return &Error{
 		op:     "umount",
 		target: target,
```

thaJeztah commented on plugin/manager_linux.go:
```diff
@@ -61,7 +61,7 @@ func (pm *Manager) enable(p *v2.Plugin, c *controller, force bool) error {
 		if p.PluginObj.Config.PropagatedMount != "" {
 			if err := mount.Unmount(propRoot); err != nil {
-				logrus.Warnf("Could not unmount %s: %v", propRoot, err)
+				logrus.Warnf("Failed to %v", err)
```
~Should this be "Failed to unmount" ?~ Ah, I see; the `.Error()` returns the "op" as part of the error, so this will include `unmount`.

Perhaps we should prefix that error with `"failed to <op>"` otherwise we'll have to add that everywhere 🤔

kolyshkin replied:
I am following the style of errors from `os` package (and the general unix error style) which is in the form `<op>: <error>`. In fact it makes sense here as well, something that would look like:

```
... level=warning msg="umount <dir>: device busy"
```

which gives no less info than
```
... level=warning msg="Failed to umount <dir>: device busy"
```

so I'll drop the "failed to" from this place.

thaJeztah commented on pkg/mount/mount.go:
```diff
@@ -2,12 +2,48 @@ package mount // import "github.com/docker/docker/pkg/mount"
+}
+
+func (e *Error) Error() string {
+	out := e.op + " "
```
Should we have the error start with "failed to " (e.g.)?

kolyshkin replied:
Well, all the error messages (say from `os` package) are in the form `op: msg`, for example

`open: permission denied`

or

`readlink: no such file or directory`

I think we should follow the pattern. Brevity is the soul of wit :smile_cat:, besides, "failed to" is redundant here -- it's an error, so it is implied that something has failed.

kolyshkin committed:
pkg/mount: refactor Unmount()

It has been pointed out that we're ignoring EINVAL from umount(2)
everywhere, so let's move it to a lower-level function. Also, its
implementation should be the same for any UNIX incarnation, so
let's consolidate it.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

thaJeztah commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1200,7 +1200,7 @@ func (devices *DeviceSet) growFS(info *devInfo) error {
 
 	if err := mount.Mount(info.DevName(), fsMountPoint, devices.BaseDeviceFilesystem, options); err != nil {
-		return fmt.Errorf("Error mounting '%s' on '%s' (fstype='%s' options='%s'): %s\n%v", info.DevName(), fsMountPoint, devices.BaseDeviceFilesystem, options, err, string(dmesg.Dmesg(256)))
+		return errors.Wrapf(err, "mount failed; dmesg: %s", string(dmesg.Dmesg(256)))
```
I think we use "failed to" in quite some locations, so from that perspective it would be good to be consistent

kolyshkin replied:
fixed some time ago (and I forgot to reply at that time)

thaJeztah commented:
Seeing one more failure;

```
18:17:27 pkg/mount/mounter_freebsd.go:1::warning: file is not goimported (goimports)

kolyshkin commented:
> Seeing one more failure

I forgot a hunk :( should be fixed now

thaJeztah reviewed (approved):
LGTM if Janky goes green

thaJeztah commented:
ping @cpuguy83 for a final look if needed

cpuguy83 commented on daemon/graphdriver/zfs/zfs.go:
```diff
@@ -390,7 +390,7 @@ func (d *Driver) Get(id, mountLabel string) (_ containerfs.ContainerFS, retErr e
 
 	if err := mount.Mount(filesystem, mountpoint, "zfs", options); err != nil {
-		return nil, fmt.Errorf("error creating zfs mount of %s to %s: %v", filesystem, mountpoint, err)
+		return nil, err
```
I think we should keep zfs in there (same with other graphdriver's error messages as well).

"error creating zfs mount" is much easier to search and find in code or to find other people reporting a similar error.

kolyshkin replied:
fixed

kolyshkin replied:
In general, though, I don't think having unique error strings to aid developers in bug report analysis is a good strategy. It would be much better to provide some context in error messages (like file name and line number where the error happens). This is the way I was doing it back in my C days (using preprocessor macros). Alas, logrus doesn't do it. Perhaps it will get this functionality one day? Could be possible with `debug.Stack()` or smth.

cpuguy83 replied:
"errors.WithStack" from "github.com/pkg/errors"... but still a good error message that signals where the error came from is important.

cpuguy83 commented on pkg/mount/mount.go:
```diff
@@ -2,12 +2,46 @@ package mount // import "github.com/docker/docker/pkg/mount"
 )
 
+// Error records an error from mount or unmount operation
+type Error struct {
```
Is there any reason to make this public?

kolyshkin replied:
I don't know, but for some reason `os.PathError` et al were made public.

I'll make it private

cpuguy83 commented on pkg/system/lstat_unix.go:
```diff
@@ -13,7 +14,7 @@ import (
 	s := &syscall.Stat_t{}
 	if err := syscall.Lstat(path, s); err != nil {
-		return nil, err
+		return nil, &os.PathError{"Lstat", path, err}
```
Can you use field names here?

cpuguy83 commented on pkg/system/stat_unix.go:
```diff
@@ -59,7 +60,7 @@ func (s StatT) IsDir() bool {
 	s := &syscall.Stat_t{}
 	if err := syscall.Stat(path, s); err != nil {
-		return nil, err
+		return nil, &os.PathError{"Stat", path, err}
```
Field names.

cpuguy83 commented on plugin/manager_linux.go:
```diff
@@ -61,7 +61,7 @@ func (pm *Manager) enable(p *v2.Plugin, c *controller, force bool) error {
 		if p.PluginObj.Config.PropagatedMount != "" {
 			if err := mount.Unmount(propRoot); err != nil {
-				logrus.Warnf("Could not unmount %s: %v", propRoot, err)
+				logrus.Warnf("Failed to %v", err)
```
```suggestion
logrus.WithField("plugin", p.GetName()).WithError(err).Warn("Error unmounting plugin propagated mount root")
```

Even the old message was bad because it's not really giving any context about the problem.

Side note, we really shouldn't assume that the error message from the mount package is following some format since it's just a string that could change and would not realize that other messages should be updated.

cpuguy83 commented on volume/local/local_unix.go:
```diff
@@ -85,8 +85,7 @@ func (v *localVolume) mount() error {
 	}
-	err := mount.Mount(v.opts.MountDevice, v.path, v.opts.MountType, mountOpts)
-	return errors.Wrapf(err, "error while mounting volume with options: %s", v.opts)
+	return mount.Mount(v.opts.MountDevice, v.path, v.opts.MountType, mountOpts)
```
Can you update the message for unmount errors in volume/service/service.go to give more context for this error?

kolyshkin replied:
In here I'm doing this:

```go
err := mount.Mount(v.opts.MountDevice, v.path, v.opts.MountType, mountOpts)
return errors.Wrap(err, "failed to mount local volume")
```

For volume/service/service.go, I don't understand how it relates to this code? Will take another look tomorrow...

cpuguy83 replied:
Sorry, this is here: https://github.com/moby/moby/blob/287144db423cc951922ee3607f024f70f2a74d25/volume/service/service.go#L106-L132

So just want to wrap the mount/unmount error at the service level so we can add details about driver name and such.

kolyshkin replied:
> So just want to wrap the mount/unmount error at the service level so we can add details about driver name and such.

@cpuguy83 from my perspective this is unrelated to this PR and can be addressed separately; this PR is already big as it is, so I'd rather merge it first.

cpuguy83 replied:
It's not exactly unrelated. We're removing these details from the built-in driver and the error messages will not mean anything unless we wrap them with appropriate context at a higher level.

kolyshkin commented:
PPC failure is https://github.com/moby/moby/issues/36903

kolyshkin commented:
Rebased; conflicts resolved.

I believe this one is ready to be merged and is very helpful is diagnosing bugs (issue of the day this PR could greatly help with is https://github.com/moby/moby/issues/38252)

kolyshkin commented:
@cpuguy83 PTAL; I know you wanted volume mount errors to be wrapped as well but I think it could be done in a separate PR -- this one is more about system mounts (as in `mount(2)`).

cpuguy83 reviewed (approved):
LGTM

thaJeztah reviewed (approved):
LGTM
